BMP581: Stream: Add FIFO Watermark Support

ubieda opened the pull request:
### Description
This PR is a follow-on effort to #93130 in which FIFO Watermark support is added on top of Streaming, and is configurable through `fifo-watermark` DTS property.

### Testing
- Build Command
```
west build -b frdm_mcxn947/mcxn947/cpu0 samples/subsys/shell/shell_module -- -DCONFIG_SENSOR=y -DCONFIG_SENSOR_SHELL=y -DCONFIG_I2C_SHELL=y -DCONFIG_SENSOR_SHELL_STREAM=y
```
- Overlay
```
#include <zephyr/dt-bindings/sensor/bmp581.h>

&flexcomm2_lpi2c2 {
	status = "okay";

	bmp581:bmp581@46 {
		compatible = "bosch,bmp581";
		reg = <0x46>;
		status = "okay";
		odr = <BMP581_DT_ODR_10_HZ>;
		press-osr = <BMP581_DT_OVERSAMPLING_1X>;
		temp-osr = <BMP581_DT_OVERSAMPLING_1X>;
		int-gpios = <&gpio0 29 (GPIO_ACTIVE_HIGH | GPIO_PULL_DOWN)>;
		fifo-watermark = <10>;
	};
};
```
- Output
```
uart:~$ sensor stream bmp581@46 on fifo_wm incl
Disabling existing stream
Enabling stream...
Trigger (10 / fifo_wm) detected
channel type=13(ambient_temp) index=0 shift=15 num_samples=10 value=60481891090ns (23.896118)
channel type=14(press) index=0 shift=15 num_samples=10 value=60481891090ns (101.374755)
Trigger (10 / fifo_wm) detected
channel type=13(ambient_temp) index=0 shift=15 num_samples=10 value=61478180456ns (23.891021)
channel type=14(press) index=0 shift=15 num_samples=10 value=61478180456ns (101.375274)
Trigger (10 / fifo_wm) detected
channel type=13(ambient_temp) index=0 shift=15 num_samples=10 value=62474455063ns (23.902160)
channel type=14(press) index=0 shift=15 num_samples=10 value=62474455063ns (101.376113)
uart:~$ sensor stream bmp581@46 off
Disabling existing stream
[00:15:21.449,000] <wrn> BMP581_STREAM: Callback triggered with no streaming submission - Disabling interrupts
```

ubieda committed:
bmp581: fix: Add parenthesis to macro arguments

Otherwise it leads to potential incorrect expansions.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

ubieda committed:
bmp581: stream: Add FIFO Watermark support

Configurable through dts property: fifo-watermark.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

ubieda committed:
bmp581: Add LOG_ERR if initialization fails

Usually, drivers failing to initialize throw a message in the log,
signaling the cause of failure. Add it to this driver so the user
isn't confused by no message yet the device being marked as disabled.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

ubieda commented:
Waiting on resolution on #93624 to decide whether to leave the closing `else {}` after every `else if {}`

fabiobaltieri commented on drivers/sensor/bosch/bmp581/bmp581_decoder.c:
```diff
@@ -105,53 +110,40 @@ static int bmp581_decoder_get_size_info(struct sensor_chan_spec chan_spec,
-
-	channel_request = bmp581_encode_channel(chan_spec.chan_type);
-	if ((channel_request & edata->header.channels) != channel_request) {
+	if (((header->events & BMP581_EVENT_FIFO_WM) != 0) && (*fit >= header->fifo_count)) {
 		return -ENODATA;
+	} else if (((header->events & BMP581_EVENT_FIFO_WM) == 0) && (*fit != 0)) {
+		return -ENODATA;
```
the two conditions are not really related right? just make two separate ifs

fabiobaltieri replied:
or, if the two conditions are related, they have the same outcome, make it a single condition

ubieda replied:
Addressed.

teburd replied:
I'm not sure I'd classify this change as "way better" at some point these boolean branch expressions spanning multiple lines become unreadable.

ubieda committed:
bmp581: code-style improvements

Remove extra-parentheses when operator precedence is implicit.
Eliminate implicit integer to boolean conversions.

Signed-off-by: Luis Ubieda <luisf@croxel.com>
